cmd/containerd-shim/main_unix.go

```diff
@@ -136,23 +139,37 @@ func serve(server *grpc.Server, path string) error {
 	return nil
 }
 
-func handleSignals(signals chan os.Signal, server *grpc.Server) error {
-	for s := range signals {
-		logrus.WithField("signal", s).Debug("received signal")
-		switch s {
-		case unix.SIGCHLD:
-			if err := reaper.Reap(); err != nil {
-				logrus.WithError(err).Error("reap exit status")
+func handleSignals(signals chan os.Signal, server *grpc.Server, sv *shim.Service) error {
+	var (
+		termOnce sync.Once
+		done     = make(chan struct{})
+	)
+
+	go func() {
+		for s := range signals {
+			switch s {
+			case unix.SIGCHLD:
+				if err := reaper.Reap(); err != nil {
+					logrus.WithError(err).Error("reap exit status")
+				}
+			case unix.SIGTERM, unix.SIGINT:
+				go termOnce.Do(func() {
+					server.Stop()
+					// Ensure our child is dead if any
+					sv.Kill(context.Background(), &shimapi.KillRequest{
+						Signal: uint32(syscall.SIGKILL),
+						All:    true,
+					})
+					sv.Delete(context.Background(), &google_protobuf.Empty{})
+					close(done)
+				})
+			case unix.SIGUSR1:
+				dumpStacks()
 			}
-		case unix.SIGTERM, unix.SIGINT:
-			// TODO: should we forward signals to the processes if they are still running?
-			// i.e. machine reboot
-			server.Stop()
-			return nil
-		case unix.SIGUSR1:
-			dumpStacks()
 		}
-	}
+	}()
+
+	<-done
```

Comment:
I need a way to to wait on `done` without waiting on the signal.

But I guess I can just use `select` there, will update